Feat/custom assertions

codefriar opened the pull request:
Adds a 'shadow' Assert class that facilitates developers adding custom assertion methods that can still be cleanly read as `Assert.methodName`

codefriar commented:
It doesn’t! That’s it’s beauty
On Oct 4, 2023 at 09:15 -0700, David Schach ***@***.***>, wrote:
> @dschach requested changes on this pull request.
> This class name requires all other Assert statements to use System.Assert? If so, it probably needs a new name.
> In force-app/main/default/classes/test utilities/Assert.cls:
> > @@ -0,0 +1,330 @@
> +/**
> + * @description Assert class 'shadows' the System.Assert class allowing us to add custom assertion methods accessed
> + * via the static Assert.* methods. This allows us to add custom assertion methods that are more readable and provide
> + * more flexibility and functionality than the standard System.Assert methods. Ultimately the goal is to provide
> + * end-developers with a more readable and flexible way to write assertions.
> + */
> ***@***.***('PMD.ExcessivePublicCount, PMD.ExcessiveParameterList')
> +public with sharing class Assert {
> If this class requires all other tests to use System.Assert, then Assert is probably not a good name for the class.
> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you authored the thread.Message ID: ***@***.***>

dschach commented:
@codefriar Does the GPT code review need to be updated? I think it may need test coverage for `Assert` class.

codefriar commented:
@dschach - I've refactored this into a new class name and reverted the shadowing work. It shows I have an outstanding request for change, but I can't find it. Thoughts on this?

dschach commented on force-app/main/default/classes/test utilities/CAssert.cls:
```diff
@@ -0,0 +1,242 @@
+	 * @param msg String - the message to display if the assertion fails
+	 */
+	public static void startsWith(String str, String prefix, String msg) {
+		Assert.isTrue(str.startsWith(prefix), msg);
```
Question: Do you want to use `startsWith` or `startsWithIgnoreCase`? I don't have an opinion - just wondering what you think.

codefriar replied:
Good question. I think principal of least surprise would dictate starts With, but maybe also a second method that's startsWithIgnoresCase ? Thoughts?

dschach commented:
Approved

dschach commented on force-app/main/default/classes/test utilities/CAssert.cls:
```diff
@@ -0,0 +1,266 @@
+	 * information, ask @jeffKrantz over on SFDC Discord.
+	 * @param msg String the message you want in your failure response
+	 */
+	public static void isNotJanky(String msg) {
+		System.Assert.fail(msg);
+	}
```
@codefriar CodeRabbit wants to know if this method is some kind of joke.
